# Airline On-Time Performance Analysis

In [1]:
import pandas as pd

In [2]:

df = pd.read_csv("../data/airline_delay_data.csv")
df.head()

,year,month,carrier,carrier_name,airport,airport_name,arr_flights,arr_del15,carrier_ct,weather_ct,...,security_ct,late_aircraft_ct,arr_cancelled,arr_diverted,arr_delay,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay
0,2026,4,G4,Allegiant Air,MCO,"Orlando, FL: Orlando International",37.0,4.0,1.00,1.00,...,0.0,0.90,1.0,0.0,238.0,38.0,107.0,49.0,0.0,44.0
1,2026,4,G4,Allegiant Air,MDT,"Harrisburg, PA: Harrisburg International",93.0,17.0,3.30,1.00,...,0.0,7.16,0.0,0.0,1723.0,1110.0,58.0,198.0,0.0,357.0
2,2026,4,G4,Allegiant Air,MDW,"Chicago, IL: Chicago Midway International",18.0,3.0,0.87,0.75,...,0.0,0.13,0.0,0.0,58.0,20.0,12.0,23.0,0.0,3.0
3,2026,4,G4,Allegiant Air,MEM,"Memphis, TN: Memphis International",42.0,16.0,3.20,1.16,...,0.0,6.14,0.0,0.0,1332.0,339.0,290.0,328.0,0.0,375.0
4,2026,4,G4,Allegiant Air,MFE,"Mission/McAllen/Edinburg, TX: McAllen Internat...",40.0,6.0,2.59,0.87,...,0.0,1.88,0.0,0.0,511.0,174.0,71.0,109.0,0.0,157.0


In [3]:
df.shape

(24400, 21)

## Initial Data Review

The dataset contains 24400 rows and 21 columns.

In [4]:
def summarize_dataframe(data):
    summary = pd.DataFrame({
        "column": data.columns,
        "data_type": data.dtypes.values,
        "missing_values": data.isnull().sum().values,
        "unique_values": data.nunique().values
    })
    
    return summary

summary_df = summarize_dataframe(df)
summary_df

,column,data_type,missing_values,unique_values
0,year,int64,0,2
1,month,int64,0,12
2,carrier,str,0,21
3,carrier_name,str,0,21
4,airport,str,0,367
5,airport_name,str,0,377
6,arr_flights,float64,7,2244
7,arr_del15,float64,14,1008
8,carrier_ct,float64,7,6103
9,weather_ct,float64,7,2048


- The dataset contains 21 columns.
- `carrier_name` contains 21 unique airlines.
-  The data includes 367 unique airport codes and 377 unique airport names.
- Most numeric columns have 7 missing values.
- `arr_del15` has the highest number of missing values, with 14 missing values.

In [5]:
df.duplicated().sum()

np.int64(0)

- 0 Duplicated records in the dataset.

In [6]:
df.describe()

,year,month,arr_flights,arr_del15,carrier_ct,weather_ct,nas_ct,security_ct,late_aircraft_ct,arr_cancelled,arr_diverted,arr_delay,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay
count,24400.000000,24400.000000,24393.000000,24386.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000,24393.000000
mean,2025.304549,6.317336,345.228713,75.780981,23.218202,2.980940,21.448760,0.142642,27.966529,6.210552,0.944615,5621.987332,1818.483458,354.911122,1185.181732,7.588693,2255.376542
std,0.460226,3.372619,993.869939,213.331968,60.758123,10.808432,73.067835,0.701058,84.520235,27.075542,4.580142,17765.149339,5952.461265,1422.725173,4819.725123,41.979275,7157.253190
min,2025.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2025.000000,4.000000,48.000000,8.000000,2.540000,0.000000,1.420000,0.000000,2.120000,0.000000,0.000000,494.000000,134.000000,0.000000,58.000000,0.000000,132.000000
50%,2025.000000,6.000000,93.000000,21.000000,6.860000,0.780000,4.540000,0.000000,7.020000,1.000000,0.000000,1418.000000,439.000000,40.000000,205.000000,0.000000,520.000000
75%,2026.000000,9.000000,232.000000,54.000000,18.540000,2.280000,12.320000,0.000000,19.180000,4.000000,1.000000,3816.000000,1324.000000,238.000000,606.000000,0.000000,1558.000000
max,2026.000000,12.000000,21854.000000,4877.000000,1334.900000,343.380000,1607.800000,29.840000,2236.130000,1463.000000,181.000000,492043.000000,170425.000000,38250.000000,182823.000000,1631.000000,240484.000000


### Descriptive Statistics Notes
- The dataset covers American flights from April 2025 to April 2026.
- The average number of arriving flights is 345.
- The median number of arriving flights is 93, which is much lower than the mean.
- This suggests that some airports or carrier-airport combinations have much higher flight volumes than others.
- The average arrival delay total is 5,622 minutes.
- Several delay columns have maximum values much higher than their medians, showing that delay totals vary widely across records.

## Data-Cleaning 

In [7]:
df_clean= df.dropna() 
df_clean.shape

(24386, 21)

After removing rows with missing values, the cleaned dataset contains 7,427 rows and 21 columns. Only 4 rows were removed, so the data loss is minimal.

In [8]:
df_clean.isnull().sum()

year                   0
month                  0
carrier                0
carrier_name           0
airport                0
airport_name           0
arr_flights            0
arr_del15              0
carrier_ct             0
weather_ct             0
nas_ct                 0
security_ct            0
late_aircraft_ct       0
arr_cancelled          0
arr_diverted           0
arr_delay              0
carrier_delay          0
weather_delay          0
nas_delay              0
security_delay         0
late_aircraft_delay    0
dtype: int64

All missing values have been removed from the cleaned dataset.

### Column Rename

In [22]:
df_clean = df_clean.rename(columns={
    "carrier_ct": "carrier_delay_count",
    "weather_ct": "weather_delay_count",
    "nas_ct": "nas_delay_count",
    "security_ct": "security_delay_count",
    "late_aircraft_ct": "lateaircraft_delay_count",
    "arr_delay": "arrival_delay_min",
    "carrier_delay": "carrier_delay_min",
    "weather_delay": "weather_delay_min",
    "arr_cancelled": "cancelled_arrivals",
    "arr_diverted": "diverted_arrivals",
    "nas_delay": "nas_delay_min",
    "security_delay": "security_delay_min",
    "late_aircraft_delay": "late_aircraft_delay_min",
    "carrier": "carrier_code",
    "airport": "airport_code",
    "arr_del15": "delayed_arrivals_15m"
})
pd.set_option("display.max_columns", None)
df_clean.head()

,year,month,carrier_code,carrier_name,airport_code,airport_name,arr_flights,delayed_arrivals_15m,carrier_delay_count,weather_delay_count,nas_delay_count,security_delay_count,lateaircraft_delay_count,cancelled_arrivals,diverted_arrivals,arrival_delay_min,carrier_delay_min,weather_delay_min,nas_delay_min,security_delay_min,late_aircraft_delay_min,delay_rate
0,2026,4,G4,Allegiant Air,MCO,"Orlando, FL: Orlando International",37.0,4.0,1.00,1.00,1.10,0.0,0.90,1.0,0.0,238.0,38.0,107.0,49.0,0.0,44.0,10.810811
1,2026,4,G4,Allegiant Air,MDT,"Harrisburg, PA: Harrisburg International",93.0,17.0,3.30,1.00,5.53,0.0,7.16,0.0,0.0,1723.0,1110.0,58.0,198.0,0.0,357.0,18.279570
2,2026,4,G4,Allegiant Air,MDW,"Chicago, IL: Chicago Midway International",18.0,3.0,0.87,0.75,1.25,0.0,0.13,0.0,0.0,58.0,20.0,12.0,23.0,0.0,3.0,16.666667
3,2026,4,G4,Allegiant Air,MEM,"Memphis, TN: Memphis International",42.0,16.0,3.20,1.16,5.51,0.0,6.14,0.0,0.0,1332.0,339.0,290.0,328.0,0.0,375.0,38.095238
4,2026,4,G4,Allegiant Air,MFE,"Mission/McAllen/Edinburg, TX: McAllen Internat...",40.0,6.0,2.59,0.87,0.66,0.0,1.88,0.0,0.0,511.0,174.0,71.0,109.0,0.0,157.0,15.000000


## Feature Engineering
Creation of a delay rate column to measure the percentage of arriving flights that were delayed by 15 minutes or more.

In [11]:
df_clean["delay_rate"] = df_clean["delayed_arrivals_15m"] / df_clean["arr_flights"] * 100

df_clean[[
    "carrier_name",
    "airport_name",
    "arr_flights",
    "delayed_arrivals_15m",
    "delay_rate"
]].head()

,carrier_name,airport_name,arr_flights,delayed_arrivals_15m,delay_rate
0,Allegiant Air,"Orlando, FL: Orlando International",37.0,4.0,10.810811
1,Allegiant Air,"Harrisburg, PA: Harrisburg International",93.0,17.0,18.279570
2,Allegiant Air,"Chicago, IL: Chicago Midway International",18.0,3.0,16.666667
3,Allegiant Air,"Memphis, TN: Memphis International",42.0,16.0,38.095238
4,Allegiant Air,"Mission/McAllen/Edinburg, TX: McAllen Internat...",40.0,6.0,15.000000


## On-Time Performance by Airline

In [16]:
on_time_performance = (
    df_clean.groupby("carrier_name")
    .agg(
        total_flights=("arr_flights", "sum"),
        delayed_flights=("delayed_arrivals_15m", "sum"),
        average_delay_rate=("delay_rate", "mean"),
        total_delay_minutes=("arrival_delay_min", "sum")
    )
)

on_time_performance["on_time_flights"] = (
    on_time_performance["total_flights"] - on_time_performance["delayed_flights"]
)

on_time_performance["otp"] = (
    on_time_performance["on_time_flights"]
    / on_time_performance["total_flights"]
    * 100
)

on_time_performance_filtered = on_time_performance[
    on_time_performance["total_flights"] >= 10000
].sort_values("otp", ascending=True)

on_time_performance_filtered

,total_flights,delayed_flights,average_delay_rate,total_delay_minutes,on_time_flights,otp
carrier_name,,,,,,
GoJet Airlines LLC d/b/a United Express,86464.0,24607.0,28.309071,2591018.0,61857.0,71.540757
Frontier Airlines,217350.0,59646.0,24.923619,5063738.0,157704.0,72.557626
JetBlue Airways,252476.0,67166.0,26.052218,5051995.0,185310.0,73.397075
PSA Airlines Inc.,264741.0,67551.0,25.850378,6306644.0,197190.0,74.484118
American Airlines Network,1064896.0,269479.0,29.127422,22668739.0,795417.0,74.694336
Allegiant Air,139718.0,34939.0,21.545643,2965630.0,104779.0,74.993201
Spirit Airlines,187759.0,46441.0,23.808955,3490339.0,141318.0,75.265633
Endeavor Air Inc.,268074.0,59561.0,23.198409,5005649.0,208513.0,77.781881
Alaska Airlines Network,297649.0,65544.0,21.022348,3404868.0,232105.0,77.979432


### Airline On-time Performance Notes
- Airlines with fewer than 10,000 total flights were excluded to avoid ranking airlines with small flight volumes in order to standardize data. 
- GoJet Airlines has the lowest on-time performance, with an OTP of 71.54%.
- Frontier Airlines has the second-lowest OTP at 72.56%, followed by JetBlue Airways at 73.40%.
- Horizon Air and Mesa Airlines notably have the highest OTP rates at 83%. 

## Delay Causes 

In [17]:
delay_causes = pd.DataFrame({
    "delay_cause": [
        "Carrier",
        "Weather",
        "NAS",
        "Security",
        "Late Aircraft"
    ],
    "delay_count": [
        df_clean["carrier_delay_count"].sum(),
        df_clean["weather_delay_count"].sum(),
        df_clean["nas_delay_count"].sum(),
        df_clean["security_delay_count"].sum(),
        df_clean["lateaircraft_delay_count"].sum()
    ]
})

delay_causes.sort_values("delay_count", ascending=False).reset_index(drop=True)

,delay_cause,delay_count
0,Late Aircraft,682187.54
1,Carrier,566361.60
2,NAS,523199.60
3,Weather,72714.07
4,Security,3479.46


In [33]:
route_delays = (
    df_clean.groupby(["airport_name", "carrier_name"])
    .agg(
        total_flights=("arrival_flights", "sum"),
        delayed_flights=("delayed_arrivals_15m", "sum"),
        total_delay_minutes=("arrival_delay_min", "sum")
    )
    .sort_values("delayed_flights", ascending=False)
    .head(5)
)

route_delays

total_flights  \
airport_name                                       carrier_name                               
Dallas/Fort Worth, TX: Dallas/Fort Worth Intern... American Airlines Network        53242.0   
Atlanta, GA: Hartsfield-Jackson Atlanta Interna... Delta Air Lines Network          76275.0   
Chicago, IL: Chicago O'Hare International          SkyWest Airlines Inc.            23639.0   
                                                   United Air Lines Network         30849.0   
Charlotte, NC: Charlotte Douglas International     American Airlines Network        36939.0   

                                                                              delayed_flights  \
airport_name                                       carrier_name                                 
Dallas/Fort Worth, TX: Dallas/Fort Worth Intern... American Airlines Network          10601.0   
Atlanta, GA: Hartsfield-Jackson Atlanta Interna... Delta Air Lines Network            10127.0   
Chicago, IL: Chicago O'Hare International          SkyWest Airlines Inc.               7634.0   
                                                   United Air Lines Network            7488.0   
Charlotte, NC: Charlotte Douglas International     American Airlines Network           5724.0   

                                                                              total_delay_minutes  
airport_name                                       carrier_name                                    
Dallas/Fort Worth, TX: Dallas/Fort Worth Intern... American Airlines Network            1079874.0  
Atlanta, GA: Hartsfield-Jackson Atlanta Interna... Delta Air Lines Network               833777.0  
Chicago, IL: Chicago O'Hare International          SkyWest Airlines Inc.                 851115.0  
                                                   United Air Lines Network              617704.0  
Charlotte, NC: Charlotte Douglas International     American Airlines Network             583587.0

## Monthly On-Time Performance

In [24]:
monthly_performance = (
    df_clean.groupby("month")
    .agg(
        total_flights=("arr_flights", "sum"),
        delayed_flights=("delayed_arrivals_15m", "sum"),
        total_delay_minutes=("arrival_delay_min", "sum")
    )
    .reset_index()
)

monthly_performance["on_time_flights"] = (
    monthly_performance["total_flights"] - monthly_performance["delayed_flights"]
)

monthly_performance["otp"] = (
    monthly_performance["on_time_flights"]
    / monthly_performance["total_flights"]
    * 100
)

monthly_performance = monthly_performance.sort_values("otp").reset_index(drop=True)

monthly_performance

,month,total_flights,delayed_flights,total_delay_minutes,on_time_flights,otp
0,7,696048.0,192888.0,15900797.0,503160.0,72.288118
1,6,674179.0,186766.0,14660004.0,487413.0,72.297268
2,12,644987.0,170133.0,13031270.0,474854.0,73.622259
3,3,675500.0,157756.0,12589161.0,517744.0,76.646040
4,5,667585.0,155262.0,11438449.0,512323.0,76.742737
5,8,666241.0,145517.0,10695249.0,520724.0,78.158504
6,11,630187.0,125230.0,9322474.0,504957.0,80.128121
7,10,668332.0,132452.0,8555392.0,535880.0,80.181706
8,1,602952.0,118672.0,9099942.0,484280.0,80.318168
9,4,1304757.0,254952.0,17567099.0,1049805.0,80.459810


### Monthly On-Time Performance Notes
- July has the lowest OTP at 72.29%, making it the worst-performing month in the dataset.
- June has a very similar OTP at 72.30%, making it the second-worst month.
- September has the highest OTP at 83.85%, making it the best-performing month.
- July also has the highest number of delayed flights, with 192,888 delayed arrivals.
- The results show weaker on-time performance during June and July compared with the other months.

In [25]:
df_clean.to_csv("../data/cleaned_airline_delay_data.csv", index=False)
on_time_performance_filtered.to_csv("../data/on_time_performance_by_airline.csv")
delay_causes.to_csv("../data/delay_causes_summary.csv", index=False)
monthly_performance.to_csv("../data/monthly_on_time_performance.csv", index=False)